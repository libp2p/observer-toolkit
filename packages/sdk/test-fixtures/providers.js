import {
  DataProvider,
  FilterProvider,
  RootNodeProvider,
} from '../components/context'

const providers = {
  DataProvider,
  FilterProvider,
  RootNodeProvider,
}

// In SDK tests, import test wrappers from '@nearform/observer-testing'
// and wrap test components with that, passing this as 'providers' prop
export default providers
