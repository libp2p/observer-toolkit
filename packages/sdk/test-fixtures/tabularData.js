import React from 'react'
import T from 'prop-types'
import { getStringSorter, getNumericSorter } from '../sorters'

const _mockState_1 = [
  { mockName: 'a', mockNumber: 3, mockBool: true, mockPercentCalc: 5 },
  { mockName: 'c', mockNumber: 2, mockBool: true, mockPercentCalc: 10 },
  { mockName: 'b', mockNumber: 4, mockBool: false, mockPercentCalc: 20 },
  { mockName: 'e', mockNumber: 1, mockBool: false, mockPercentCalc: 14 },
  { mockName: 'd', mockNumber: 5, mockBool: true, mockPercentCalc: 1 },
]

const _mockState_0 = [
  Object.assign({}, _mockState_1[4]),
  Object.assign({}, _mockState_1[3]),
  Object.assign({}, _mockState_1[2]),
  Object.assign({}, _mockState_1[1], { mockNumber: 0 }),
  Object.assign({}, _mockState_1[0]),
]

const mockStates = [_mockState_0, _mockState_1]

const _MockBoolRenderer = function({ boolProp, value }) {
  return boolProp ? <em>{value}</em> : <strong role="alert">{value}</strong>
}
_MockBoolRenderer.propTypes = {
  boolProp: T.bool.isRequired,
  value: T.string.isRequired,
}

const stringSorter = {
  getSorter: getStringSorter,
  defaultDirection: 'asc',
}

const numericSorter = {
  getSorter: getNumericSorter,
  defaultDirection: 'desc',
}

const mockColumnDefs = [
  {
    name: 'mockName',
    sort: stringSorter,
    header: <strong>Name of item</strong>,
    rowKey: 'value',
  },
  {
    name: 'mockNumber',
    sort: numericSorter,
    align: 'right',
    header: 'Numeric cells',
  },
  {
    name: 'mockBool',
    getProps: ({ mockBool }) => ({
      value: mockBool ? 'okay' : 'ALERT!',
      boolProp: mockBool,
    }),
    renderContent: _MockBoolRenderer,
  },
  {
    name: 'mockPercentCalc',
    getProps: ({ mockPercentCalc }, metadata) => {
      const percent = Math.round(
        (mockPercentCalc / metadata.percentTotal) * 100
      )
      return {
        value: `${percent}%`,
        sortValue: mockPercentCalc,
      }
    },
    sort: numericSorter,
  },
]

export { mockStates, mockColumnDefs }
