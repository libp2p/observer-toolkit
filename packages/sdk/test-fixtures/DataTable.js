import React, { useContext, useState } from 'react'

import { ThemeWrapper } from '@libp2p-observer/testing'

import { mockStates, mockColumnDefs } from './tabularData'
import providers from './providers'
import { useTabularData } from '../hooks'
import { DataTable, StyledButton } from '../components'

function MockDataTable() {
  const [data, setData] = useState(mockStates[mockStates.length - 1])

  const metadata = {
    percentTotal: data.reduce(
      (total, { mockPercentCalc }) => total + mockPercentCalc,
      0
    ),
  }

  const {
    columnDefs,
    contentProps,
    sortColumn,
    setSortColumn,
    sortDirection,
    setSortDirection,
  } = useTabularData({
    columns: mockColumnDefs,
    data,
    defaultSort: 'mockName',
    metadata,
  })

  return (
    <div>
      <div data-testid="data-switchers">
        {mockStates.map((datum, index) => (
          <StyledButton
            key={`button_${index}`}
            onClick={() => setData(mockStates[index])}
          >
            Select [{index}]
          </StyledButton>
        ))}
      </div>
      <DataTable
        contentProps={contentProps}
        columnDefs={columnDefs}
        sortColumn={sortColumn}
        setSortColumn={setSortColumn}
        sortDirection={sortDirection}
        setSortDirection={setSortDirection}
      />
    </div>
  )
}

function ThemedMockDataTable() {
  return (
    <ThemeWrapper providers={providers}>
      <MockDataTable />
    </ThemeWrapper>
  )
}

export { MockDataTable, ThemedMockDataTable }
