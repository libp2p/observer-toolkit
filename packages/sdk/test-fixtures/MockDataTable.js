import React, { useState } from 'react'
import T from 'prop-types'

import { ThemeWrapper } from '@libp2p/observer-testing'

import { mockStates, mockColumnDefs } from './tabularData'
import providers from './providers'
import { useTabularData } from '../hooks'
import { DataTable, StyledButton } from '../components'

function MockDataTable({ states = mockStates, hasPagination = false }) {
  const [data, setData] = useState(states[states.length - 1])

  const metadata = {
    percentTotal: data.reduce(
      (total, { mockPercentCalc }) => total + mockPercentCalc,
      0
    ),
  }

  const rowsPerPageOptions = [2, 3, 4, 5, 10]
  const defaultPerPageIndex = 3
  const range = hasPagination
    ? [0, rowsPerPageOptions[defaultPerPageIndex]]
    : null

  const {
    allContent,
    shownContent,
    columnDefs,
    sortColumn,
    setSortColumn,
    sortDirection,
    setSortDirection,
    rowCounts,
    setRange,
  } = useTabularData({
    columns: mockColumnDefs,
    data,
    defaultSort: 'mockName',
    metadata,
    range,
  })

  const paginationProps = hasPagination
    ? {
        hasPagination,
        rowsPerPageOptions,
        defaultPerPageIndex,
        setRange,
      }
    : {}

  return (
    <div>
      <div data-testid="data-switchers">
        {states.map((datum, index) => (
          <StyledButton
            key={`button_${index}`}
            onClick={() => setData(states[index])}
            isActive={states[index] === data}
          >
            Select [{index}]
          </StyledButton>
        ))}
      </div>
      <DataTable
        allContent={allContent}
        shownContent={shownContent}
        columnDefs={columnDefs}
        sortColumn={sortColumn}
        setSortColumn={setSortColumn}
        sortDirection={sortDirection}
        setSortDirection={setSortDirection}
        rowCounts={rowCounts}
        {...paginationProps}
      />
    </div>
  )
}
MockDataTable.propTypes = {
  states: T.array,
  hasPagination: T.bool,
}

function ThemedMockDataTable(props) {
  return (
    <ThemeWrapper providers={providers}>
      <MockDataTable {...props} />
    </ThemeWrapper>
  )
}

export { MockDataTable, ThemedMockDataTable }
