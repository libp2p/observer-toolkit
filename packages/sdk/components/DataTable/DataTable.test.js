import React from 'react'
import { act, fireEvent, wait } from '@testing-library/react'
import { renderForSDK, within } from '@libp2p-observer/testing'

import { ThemedMockDataTable } from '../../test-fixtures/MockDataTable'

function getSlidingRows(nodes) {
  const slidingRows = nodes.reduce((rows, node) => {
    const row = node.querySelectorAll(':scope > table tr')
    return row.length === 1 ? [...rows, ...row] : rows
  }, [])
  return slidingRows
}

describe('DataTable', () => {
  it('Renders as expected, matching snapshot', () => {
    const { asFragment } = renderForSDK(<ThemedMockDataTable />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('Renders cell contents correctly', () => {
    const { getByTableRow } = renderForSDK(<ThemedMockDataTable />)

    // Find by name cell
    const rowA = getByTableRow([{ column: 'Name of item', textContent: 'a' }])
    const rowA_numeric = within(rowA).getByTableColumn('Numeric cells')

    expect(rowA_numeric.textContent).toBe('3')

    const rowA_bool = within(rowA).getByTableColumn('mockBool')
    expect(rowA_bool.textContent).toBe('okay')
    expect(within(rowA_bool).queryAllByRole('alert')).toHaveLength(0)

    const rowA_percent = within(rowA).getByTableColumn('mockPercentCalc')
    expect(rowA_percent.textContent).toBe('10%')

    // Find by number cell
    const rowB = getByTableRow([{ column: 'Numeric cells', numericContent: 4 }])
    const rowB_name = within(rowB).getByTableColumn('Name of item')
    expect(rowB_name.textContent).toBe('b')

    const rowB_bool = within(rowB).getByTableColumn('mockBool')
    expect(rowB_bool.textContent).toBe('ALERT!')
    expect(within(rowB_bool).queryAllByRole('alert')).toHaveLength(1)

    const rowB_percent = within(rowB).getByTableColumn('mockPercentCalc')
    expect(rowB_percent.textContent).toBe('40%')
  })

  it('Renders cells in correct default order', () => {
    const { getAllByTableColumn } = renderForSDK(<ThemedMockDataTable />)

    const nameCol = getAllByTableColumn('Name of item')
    const nameColContent = nameCol.map(cell => cell.textContent)
    expect(nameColContent).toEqual(['a', 'b', 'c', 'd', 'e'])

    const numCol = getAllByTableColumn('Numeric cells')
    const numColContent = numCol.map(cell => Number(cell.textContent))
    expect(numColContent).toEqual([3, 4, 2, 5, 1])
  })

  it('Changes cell order on clicking sort column headers', async () => {
    const {
      findAllByRole,
      getByText,
      getAllByTableColumn,
      queryAllByRole,
    } = renderForSDK(<ThemedMockDataTable />)

    await act(async () => {
      await fireEvent.click(getByText('Numeric cells'))
    })
    // rows except 'b' and 'e' should slide
    expect(getSlidingRows(await findAllByRole('presentation'))).toHaveLength(3)
    expect(queryAllByRole('table')).toHaveLength(4)

    // wait for sliding to complete and sliding rows to be erased
    await wait(() => {
      expect(queryAllByRole('table')).toHaveLength(1)
    })

    const nameCol = getAllByTableColumn('Name of item')
    const nameColContent = nameCol.map(cell => cell.textContent)
    expect(nameColContent).toEqual(['d', 'b', 'a', 'c', 'e'])

    const numCol = getAllByTableColumn('Numeric cells')
    const numColContent = numCol.map(cell => Number(cell.textContent))
    expect(numColContent).toEqual([5, 4, 3, 2, 1])
  })

  it('Sort persists and updates as datum changes', async () => {
    const {
      findAllByRole,
      getByText,
      queryAllByRole,
      queryByRole,
    } = renderForSDK(<ThemedMockDataTable />)

    await act(async () => {
      await fireEvent.click(getByText('Numeric cells'))
    })

    await wait(() => {
      expect(queryAllByRole('table')).toHaveLength(1)
    })

    await act(async () => {
      await fireEvent.click(getByText('Select [0]'))
    })

    expect(getSlidingRows(await findAllByRole('presentation'))).toHaveLength(2)

    await wait(() => {
      expect(queryAllByRole('table')).toHaveLength(1)
    })

    const table = queryByRole('table')

    const nameCol = within(table).getAllByTableColumn('Name of item')
    const nameColContent = nameCol.map(cell => cell.textContent)
    expect(nameColContent).toEqual(['d', 'b', 'a', 'e', 'c'])

    const numCol = within(table).getAllByTableColumn('Numeric cells')
    const numColContent = numCol.map(cell => Number(cell.textContent))
    expect(numColContent).toEqual([5, 4, 3, 1, 0])
  })
})
