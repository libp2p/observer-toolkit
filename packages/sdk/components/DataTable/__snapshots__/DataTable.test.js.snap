// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable Renders as expected, matching snapshot 1`] = `
<DocumentFragment>
  .c0 {
  width: inherit;
  height: inherit;
  display: inherit;
  -webkit-flex-direction: inherit;
  -ms-flex-direction: inherit;
  flex-direction: inherit;
}

.c7 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  vertical-align: middle;
  width: 2em;
  height: 2em;
}

.c10 {
  display: inline-block;
}

.c1 {
  padding: 4px 8px;
  position: relative;
  z-index: 5;
  cursor: pointer;
  font-weight: bold;
  white-space: nowrap;
  background: rgb(255,255,255);
  border: 1px solid rgb(209,14,102);
  color: rgb(209,14,102);
}

.c1:focus {
  outline: none;
}

.c3 {
  border-collapse: collapse;
  table-layout: fixed;
  width: 100%;
}

.c11 {
  border-bottom: 1px solid rgba(241,242,240,0.5);
}

.c4 {
  font-family: plex-sans,sans-serif;
  line-height: 1.6em;
  font-size: 9pt;
  line-height: 1.2em;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 24px;
  padding-right: 16px;
  text-align: left;
  padding-right: 0;
  font-family: plex-sans;
  font-weight: 600;
  background-color: rgba(32,155,198,0.2);
  color: rgba(93,94,96,0.8);
  text-transform: uppercase;
  white-space: nowrap;
}

.c8 {
  font-family: plex-sans,sans-serif;
  line-height: 1.6em;
  font-size: 9pt;
  line-height: 1.2em;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 24px;
  padding-right: 16px;
  text-align: left;
  padding-right: 0;
  font-family: plex-sans;
  font-weight: 600;
  background-color: rgba(32,155,198,0.2);
  color: rgba(93,94,96,0.8);
  text-transform: uppercase;
  white-space: nowrap;
  text-align: right;
}

.c12 {
  font-family: plex-sans,sans-serif;
  line-height: 1.6em;
  font-size: 9pt;
  line-height: 1.2em;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 24px;
  padding-right: 16px;
  text-align: left;
  padding-right: 32px;
  font-weight: 400;
  font-family: plex-sans;
  color: rgb(93,94,96);
}

.c13 {
  font-family: plex-sans,sans-serif;
  line-height: 1.6em;
  font-size: 9pt;
  line-height: 1.2em;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 24px;
  padding-right: 16px;
  text-align: left;
  padding-right: 32px;
  font-weight: 400;
  font-family: plex-sans;
  color: rgb(93,94,96);
  text-align: right;
}

.c6 {
  height: 2em;
  width: 2em;
  display: inline-block;
  vertical-align: middle;
  margin-left: 4px;
  -webkit-transition: -webkit-transform 0.4s ease-in-out;
  -webkit-transition: transform 0.4s ease-in-out;
  transition: transform 0.4s ease-in-out;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.c9 {
  height: 2em;
  width: 2em;
  display: inline-block;
  vertical-align: middle;
  margin-left: 4px;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.c5 {
  cursor: pointer;
}

.c2 {
  position: relative;
}

<div
    class="c0"
  >
    <div>
      <div
        data-testid="data-switchers"
      >
        <button
          class="c1"
        >
          Select [0]
        </button>
        <button
          class="c1"
        >
          Select [1]
        </button>
      </div>
      <div
        class="c2"
      >
        <table
          class="c3"
        >
          <thead
            class=""
          >
            <tr
              class=""
            >
              <th
                class="c4"
              >
                <div
                  class="c5"
                >
                  <strong>
                    Name of item
                  </strong>
                  <span
                    class="c6"
                  >
                    <span
                      class="c7"
                      role="img"
                    >
                      <svg
                        viewBox="0 0 32 32"
                      >
                        <path
                          d="M18.933 24.716a.85.85 0 00.601-1.451l-7.264-7.263 7.264-7.267a.85.85 0 00-1.201-1.202l-7.866 7.868a.852.852 0 00.001 1.203l7.865 7.864a.846.846 0 00.6.248z"
                          fill="currentColor"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </th>
              <th
                class="c8"
              >
                <div
                  class="c5"
                >
                  Numeric cells
                  <span
                    class="c9"
                  >
                    <span
                      class="c7"
                      role="img"
                    >
                      <span
                        class="c10"
                      />
                    </span>
                  </span>
                </div>
              </th>
              <th
                class="c4"
              >
                <div
                  class=""
                >
                  mockBool
                </div>
              </th>
              <th
                class="c4"
              >
                <div
                  class="c5"
                >
                  mockPercentCalc
                  <span
                    class="c9"
                  >
                    <span
                      class="c7"
                      role="img"
                    >
                      <span
                        class="c10"
                      />
                    </span>
                  </span>
                </div>
              </th>
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="c11"
            >
              <td
                class="c12"
              >
                a
              </td>
              <td
                class="c13"
              >
                3
              </td>
              <td
                class="c12"
              >
                <em>
                  okay
                </em>
              </td>
              <td
                class="c12"
              >
                10%
              </td>
            </tr>
            <tr
              class="c11"
            >
              <td
                class="c12"
              >
                b
              </td>
              <td
                class="c13"
              >
                4
              </td>
              <td
                class="c12"
              >
                <strong
                  role="alert"
                >
                  ALERT!
                </strong>
              </td>
              <td
                class="c12"
              >
                40%
              </td>
            </tr>
            <tr
              class="c11"
            >
              <td
                class="c12"
              >
                c
              </td>
              <td
                class="c13"
              >
                2
              </td>
              <td
                class="c12"
              >
                <em>
                  okay
                </em>
              </td>
              <td
                class="c12"
              >
                20%
              </td>
            </tr>
            <tr
              class="c11"
            >
              <td
                class="c12"
              >
                d
              </td>
              <td
                class="c13"
              >
                5
              </td>
              <td
                class="c12"
              >
                <em>
                  okay
                </em>
              </td>
              <td
                class="c12"
              >
                2%
              </td>
            </tr>
            <tr
              class="c11"
            >
              <td
                class="c12"
              >
                e
              </td>
              <td
                class="c13"
              >
                1
              </td>
              <td
                class="c12"
              >
                <strong
                  role="alert"
                >
                  ALERT!
                </strong>
              </td>
              <td
                class="c12"
              >
                28%
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
